### PR TITLE
Redirect loop when URL have uppercase

### LINKED ISSUE #38538: Redirect loop when URL have uppercase
### Preconditions and environment

- Magento version 2.4-develop

### Steps to reproduce

1. Run: php bin/magento con:set web/secure/base_url https://Magento2.test/ ( or change Secure Base URL value in admin)
1. Clean cache: php bin/magento cache:clean
1. Go to homepage

### Expected result

URL should convert to lowercase

### Actual result

Redirect loop
!https://github.com/magento/magento2/assets/78783748/90e91b39-6f89-49c3-99c5-a2d2ebba718c!

### Additional information

*No response*

### Release note

*No response*

### Triage and priority

- < ] Severity: **S0** *- Affects critical data or functionality and leaves users without workaround.*
- [ ] Severity: **S1** *- Affects critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S2** *- Affects non-critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S3** *- Affects non-critical data or functionality and does not force users to employ a workaround.*
- [ > Severity: **S4** *- Affects aesthetics, professional look and feel, "quality" or "usability".*

### PR DESCRIPTION
**Pull request by rogerdz**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#38538

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by rogerdz** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @rogerdz,

Thanks for the contribution!

Please fix the failed tests.

Thanks

**Comment by engcom-Dash** [comment]
@magento give me test instance

**Comment by engcom-Dash** [comment]
Fixed failing unit tests so moving this back to pending review.

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run all tests

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php [hunk 1]
```diff
@@ -61,4 +69,47 @@ public function testSaveMergedJsCssMustBeCleaned()
...
         $model->afterSave();
     }
+
+    /**
+     * @dataProvider beforeSaveDataProvider
+     * @param string $value
+     * @param string $expectedValue
+     * @return void
+     */
+    public function testBeforeSaveConvertLowerCase($value, $expectedValue)
```
Comment (engcom-Hotel) [inline app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php]: Add type hints (string for parameters and void for return type)

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php [hunk 2]
```diff
@@ -61,4 +69,47 @@ public function testSaveMergedJsCssMustBeCleaned()
         $model->setValue('http://example.com/')->setPath(Store::XML_PATH_UNSECURE_BASE_URL);
         $model->afterSave();
     }
+
+    /**
+     * @dataProvider beforeSaveDataProvider
```
Comment (engcom-Hotel) [inline app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:76]: Add a detailed DocBlock for the method.

for e.g.
```
/**
 * Test beforeSave method to ensure URL is converted to lower case.
 *
 * @dataProvider beforeSaveDataProvider
 * @param string $value
 * @param string $expectedValue
 * @return void
 */
```

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php [hunk 3]
```diff
@@ -61,4 +69,47 @@ public function testSaveMergedJsCssMustBeCleaned()
...
+
+        $model->setValue($value);
+        $model->beforeSave();
+        $this->assertEquals($expectedValue, $model->getValue());
+    }
+
+    /**
+     * @return array
+     */
+    public static function beforeSaveDataProvider()
```
Comment (engcom-Hotel) [inline app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php]: Add a return type hint (array)

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php [hunk 4]
```diff
@@ -61,4 +69,47 @@ public function testSaveMergedJsCssMustBeCleaned()
...
+            ->with($expectedValue, ['http', 'https'])
+            ->willReturn(true);
+
+        $model->setValue($value);
+        $model->beforeSave();
+        $this->assertEquals($expectedValue, $model->getValue());
+    }
+
+    /**
+     * @return array
```
Comment (engcom-Hotel) [inline app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:109]: Add a detailed DocBlock for the method.

```
/**
 * Data provider for testBeforeSaveConvertLowerCase.
 *
 * @return array
 */
```

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @engcom-Dash,

Please fix the below review comments as well.

Thanks

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php (hunk 4 above)
Comment (engcom-Dash) [reply app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:109]: @engcom-Hotel added detailed doc block.

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php (hunk 2 above)
Comment (engcom-Dash) [reply app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:76]: @engcom-Hotel added detailed doc block.

**Comment by engcom-Dash** [comment]
@magento run all tests

**Review by engcom-Hotel — APPROVED** [review]

**Comment by engcom-Bravo** [comment]
Hi @rogerdz,

Thanks for the collaboration & contribution!

### :heavy_check_mark: QA Passed


**Preconditions:**

- Install fresh Magento **`2.4-develop`**

**Steps to reproduce**

- Run: php bin/magento con:set web/secure/base_url https://Magento2.test/ ( or change Secure Base URL value in admin)
- Clean cache: php bin/magento cache:clean
- Go to homepage


**Before:** :heavy_multiplication_x: 

Screenshot 2024-09-24 at 11 35 32




**After:** :heavy_check_mark: 


Screenshot 2024-09-24 at 11 48 43




Builds are failed. Hence, moving this PR to **Extended Testing**.

Thanks

File: app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php
Comment (hostep) [inline app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:1]: Please revert back the file permissions on this file from 0755 to 0644, there's no good reason to change them to 0755 ...

Comment (engcom-Dash) [reply app/code/Magento/Config/Test/Unit/Model/Config/Backend/BaseurlTest.php:1]: Hello @hostep, Reverted the file permissions.

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Dash** [comment]
The Functional B2B and EE test failures are not consistent in recent 2 builds. Seems to be flaky, hence moving it to Merge in Progress.
Functional B2B:
Run 1:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38539/97fba4c05927ed773c07ac5966214654/Functional/allure-report-b2b/index.html#categories/16b25a50b4a8b68f5548cdba82ebd5f8/ce6350f35f68e89d/
image

Run 2:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38539/30a6bc498427a4b59e1324aa693f15c4/Functional/allure-report-b2b/index.html#categories/39002bb4b823f2d59e22404a186a7115/d3d40d3f286a0c58/
image

Functional EE
Run 1: 
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38539/41e1c14f0e05ca541e402e6c731ced2f/Functional/allure-report-ee/index.html#categories/75b40a63f79830217aae704c6d6af940/de4531c7ce72aa00/
image

Run 2:
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/38539/526ad0af24633476a36188d5f05e41bd/Functional/allure-report-ee/index.html#categories/75b40a63f79830217aae704c6d6af940/d3bd81128df7f8a5/
image

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Pull request merged**